chain/client/src/chunk_inclusion_tracker.rs

```diff
@@ -0,0 +1,100 @@
+use chrono::{DateTime, Utc};
+use itertools::Itertools;
+use lru::LruCache;
+use std::collections::HashMap;
+
+use near_primitives::hash::CryptoHash;
+use near_primitives::sharding::ShardChunkHeader;
+use near_primitives::types::{AccountId, EpochId, ShardId};
+
+use crate::metrics;
+
+const CHUNK_HEADERS_FOR_INCLUSION_CACHE_SIZE: usize = 2048;
+const NUM_EPOCH_CHUNK_PRODUCERS_TO_KEEP_IN_BLOCKLIST: usize = 1000;
+
+struct ChunkInfo {
+    pub chunk_header: ShardChunkHeader,
+    pub received_time: DateTime<Utc>,
+    pub chunk_producer: AccountId,
+}
+pub struct ChunkInclusionTracker {
+    prev_block_to_chunk_headers_ready_for_inclusion: LruCache<
+        CryptoHash,
+        HashMap<ShardId, (ShardChunkHeader, chrono::DateTime<chrono::Utc>, AccountId)>,
+    >,
+
+    banned_chunk_producers: LruCache<(EpochId, AccountId), ()>,
+}
+
+impl ChunkInclusionTracker {
+    pub fn new() -> Self {
+        Self {
+            prev_block_to_chunk_headers_ready_for_inclusion: LruCache::new(
+                CHUNK_HEADERS_FOR_INCLUSION_CACHE_SIZE,
+            ),
+            // chunk_header_info: LruCache::new(CHUNK_HEADERS_FOR_INCLUSION_CACHE_SIZE * 10),
+            banned_chunk_producers: LruCache::new(NUM_EPOCH_CHUNK_PRODUCERS_TO_KEEP_IN_BLOCKLIST),
+        }
+    }
+
+    pub fn mark_chunk_header_ready_for_inclusion(
+        &mut self,
+        chunk_header: ShardChunkHeader,
+        chunk_producer: AccountId,
+    ) {
+        let prev_block_hash = chunk_header.prev_block_hash();
+        self.prev_block_to_chunk_headers_ready_for_inclusion
+            .get_or_insert(*prev_block_hash, || HashMap::new());
+        self.prev_block_to_chunk_headers_ready_for_inclusion
+            .get_mut(prev_block_hash)
+            .unwrap()
```

Comment:
While you're at it, do you want to clean this up to entry.or_default?

Reply:
We can't... The LRU cache doesn't have our nice entry interface :(